walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -13,11 +13,157 @@
 // limitations under the License.
 
 package com.google.sps;
-
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.ArrayList;
 import java.util.Collection;
 
+/**
+ * Class which finds a range of times when all required attendees can meet
+ * If there are no available times when an all optional attendees can meet,
+ * then only required attendees are considered in finding available times 
+ */
 public final class FindMeetingQuery {
+
+  /**
+   * Find available time ranges to meet
+   * @param events All events which are occuring 
+   * @param request A requested meeting with certain required and optional attendees and a duration
+   * @return all time ranges when the required attendess, and possibly the optional attendees, can meet
+   */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Collection<String> allAttendees = new ArrayList<String>();
+    allAttendees.addAll(request.getAttendees());
+    allAttendees.addAll(request.getOptionalAttendees());
+    Collection<TimeRange> openTimes = queryWithAttendees(events, request, allAttendees);
+    if (request.getAttendees().size() == 0 || openTimes.size() != 0)
+      return openTimes;
+    return queryWithAttendees(events, request, request.getAttendees());
+  }
+
+  /**
+   * Find available time ranges to meet for a specific collection of attendees
+   * @param events All events which are occuring 
+   * @param request A requested meeting with certain required and optional attendees and a duration
+   * @param attendees A collection of attendees which will be considered in finding available time ranges
+   * @return all time ranges when the {@code attendess} can meet
+   */
+  private Collection<TimeRange> queryWithAttendees(Collection<Event> events,
+                                                  MeetingRequest request,
+                                                  Collection<String> attendees) {
+    Collection<Event> restrainingEvents = filterEventsByAttendees(events, attendees);
+    Collection<TimeRange> eventTimeRanges = getTimeRangesFromEvents(restrainingEvents);
+    Collection<TimeRange> occupiedTimes = getNonOverlappingTimes(eventTimeRanges);
+    return getAvailableTimes(occupiedTimes, request.getDuration());
+  }
+
+  /** 
+   * Checks if time overlaps with addedTimes. If it does, it will change addedTimes so that
+   * addedTimes contains time in it without having any overlap
+   * @param addedTimes All the {@code TimeRange} which do not overlap
+   * @param time A time to be added to {@code addedTimes}. {@code addedTimes} is changed to include time
+   * without overlapping any {@code TimeRange}
+   */
+  private void mergeToAddedTimes(Collection<TimeRange> addedTimes, TimeRange time) {
+    Collection<TimeRange> overlappingTimes = new ArrayList<TimeRange>();
```

Comment:
optional nit: I think the added times naming here is a bit confusing, maybe 
mergeTimeWithCollection(Collection<timeRange> existingTimes, TimeRange time) or something similar?